### PR TITLE
test(language): Add comprehensive AST visitor and printer robustness tests

### PR DESCRIPTION
**Pull request by pekkah**
## Summary
- Add 25+ comprehensive robustness tests for AST visitor and printer components
- Test edge cases, boundary conditions, and error scenarios for document traversal and printing
- Verify performance characteristics and memory efficiency under stress conditions
- Include round-trip testing to ensure parse-print consistency

## Changes Made
- **New test file**: `ASTVisitorAndPrinterRobustnessFacts.cs` with comprehensive coverage:
  - **Document Walker Robustness**: Null inputs, deep nesting, large documents, efficiency testing
  - **Visitor Exception Handling**: Error propagation, multiple visitor coordination
  - **Printer Robustness**: Null inputs, unicode content, special characters, memory efficiency
  - **Printer Context Testing**: Utility methods (rewind, endsWith), large output handling
  - **Round-trip Testing**: Parse-print-parse consistency verification
  - **Performance Testing**: Memory leak prevention, consistent performance

## Test Categories
1. **Document Walker Robustness** (5 tests): Edge cases for AST traversal
2. **Visitor Exception Handling** (3 tests): Error handling and multi-visitor scenarios
3. **Printer Robustness** (8 tests): Various input edge cases and boundary conditions
4. **Printer Context Edge Cases** (5 tests): Utility method testing and large output handling
5. **Round-trip Testing** (2 tests): Parse-print consistency verification
6. **Performance and Memory Tests** (2 tests): Efficiency and memory leak prevention

## Key Robustness Features
- **Null Safety**: Tests with null visitors, documents, and inputs
- **Deep Nesting**: Stress tests with 100+ levels of nested structures
- **Large Documents**: Performance tests with 1000+ field documents
- **Unicode Support**: Proper handling of emoji and international characters
- **Memory Efficiency**: Tests for memory leaks and performance consistency
- **Error Resilience**: Exception handling and graceful degradation

## Coverage Areas
The tests verify robustness for:
- AST document walker with various edge case inputs
- Visitor pattern with exception handling and multiple visitors
- Printer output generation with complex nested structures
- Printer context utility operations and memory management
- Round-trip consistency between parsing and printing
- Performance characteristics under load conditions

## Part of Test Coverage Improvement Plan
This is **PR 3 of 15** in the systematic test coverage improvement plan for issue #1977.

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: tests/GraphQL.Language.Tests/ASTVisitorAndPrinterRobustnessFacts.cs [hunk 1]
```diff
@@ -0,0 +1,541 @@
...
+        var document = Parser.Create("{ field }").ParseExecutableDocument();
+        var visitors = new List<IReadOnlyDocumentVisitor<PrinterContext>?> { null };
+        
+        // When & Then: Should handle null visitor gracefully
+        var context = new PrinterContext();
+        var exception = Record.Exception(() => 
+        {
+            var walker = new ReadOnlyDocumentWalker<PrinterContext>(
+                visitors.Where(v => v != null), 
+                context
```
[the remark this answers is not in the document]
Comment (pekkah) [reply tests/GraphQL.Language.Tests/ASTVisitorAndPrinterRobustnessFacts.cs]: Fixed! Added `.ToList()` to convert the `IEnumerable<T>` result from `Where()` to `IReadOnlyList<T>` as expected by the constructor. Also reconsidered the test - it now properly tests filtering null visitors before walker creation.

File: tests/GraphQL.Language.Tests/ASTVisitorAndPrinterRobustnessFacts.cs [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+    {
+        // Given: Null node
+        INode nullNode = null!;
+
+        // When & Then: Should handle null input gracefully
+        var exception = Record.Exception(() => Printer.Print(nullNode));
+        
+        // Depending on implementation, may throw or return empty string
+        // This test documents the current behavior
+        Assert.NotNull(exception);
```
[the remark this answers is not in the document]
Comment (pekkah) [reply tests/GraphQL.Language.Tests/ASTVisitorAndPrinterRobustnessFacts.cs]: Good catch! Changed the assertion to `Assert.Null(exception)` since `Printer.Print` uses `ReadOnlyDocumentWalker`, which handles null nodes by returning early without throwing an exception. The printer should handle null nodes gracefully.

File: tests/GraphQL.Language.Tests/ASTVisitorAndPrinterRobustnessFacts.cs [hunk 3]
```diff
@@ -0,0 +1,541 @@
...
+        // Given: Document with special characters that need escaping
+        var source = @"{ field(arg: ""line1\nline2\ttab\r\n"") }";
+        var document = Parser.Create(source).ParseExecutableDocument();
+
+        // When: Print the document
+        var result = Printer.Print(document);
+
+        // Then: Should handle escape sequences properly
+        Assert.NotNull(result);
+        Assert.Contains("field", result);
```
[the remark this answers is not in the document]
Comment (pekkah) [reply tests/GraphQL.Language.Tests/ASTVisitorAndPrinterRobustnessFacts.cs:330]: Good suggestion! Enhanced the test with a more specific assertion `Assert.Contains("\"\"\"", result)` to verify that the printer correctly uses block string quotes for strings containing newlines, making the test more robust.

File: tests/GraphQL.Language.Tests/ASTVisitorAndPrinterRobustnessFacts.cs [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+    {
+        // Given: Document that will produce large output
+        var largeQuery = "{ ";
+        for (int i = 0; i < 1000; i++)
+        {
+            largeQuery += $"field{i} ";
+        }
+        largeQuery += "}";
+        
+        var document = Parser.Create(largeQuery).ParseExecutableDocument();
```
[the remark this answers is not in the document]
Comment (pekkah) [reply tests/GraphQL.Language.Tests/ASTVisitorAndPrinterRobustnessFacts.cs]: Absolutely right! Replaced the inefficient string concatenation with `StringBuilder` for better performance when generating the large query with 1000 fields.

File: tests/GraphQL.Language.Tests/ASTVisitorAndPrinterRobustnessFacts.cs [hunk 5]
```diff
@@ -0,0 +1,541 @@
...
+        {
+            // When: Parse then print then parse again
+            var document = Parser.Create(originalSource).ParseExecutableDocument();
+            var printed = Printer.Print(document);
+            
+            // Then: Should produce valid GraphQL that can be parsed
+            var exception = Record.Exception(() => Parser.Create(printed).ParseExecutableDocument());
+            Assert.Null(exception);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (pekkah) [reply tests/GraphQL.Language.Tests/ASTVisitorAndPrinterRobustnessFacts.cs]: Good call! Removed the redundant `Printer_SimpleRoundtrip_WorksCorrectly` test as its logic was already covered by the more comprehensive `Printer_ParsePrintRoundtrip_MaintainsEquivalence` test. This improves maintainability and avoids duplicated test logic. Note: There are still some compilation errors being addressed due to API changes in the ReadOnlyDocumentWalker generics.
